main.cpp

```diff
@@ -35,74 +35,66 @@ struct T
 
 struct X                                //4
 {
-    T* compare(T* a, T* b) //5
+    const T* compare(const T & a, const T & b) //5
```

Comment:
You're the first person to recognize that making the arg type `const T&` forces the return type to be `const T*`, not `T*`.   nicely done.    most people change the parameter type from `T* to `T&`